### PR TITLE
Create link to map on higlass server and don't duplicate files on higlass server

### PR DESCRIPTION
**Pull request by BethYates**
Check if data exists already exists on the Higlass server before adding it.Removes file if already present before re-uploading, outputs uuid of uploaded files for use when creating link to higlass server.  Generates link to the uploaded higlass map. 

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/sanger-tol/genomenote/tree/master/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by BethYates** [comment]
A new module was added to generate the link to the higlass map. It uses the python request module to make a post request to the higlass API with the specified viewconfig. This request returns a uid for the view config which can then be used to form the link to the map.

**Comment by gq1** [comment]
Nothing to do with this PR, just wondering why the Github CI test being removed on Pull Request? I can't see any CI Test being run on Docker profile for this PR.

**Comment by BethYates** [comment]
> Nothing to do with this PR, just wondering why the Github CI test being removed on Pull Request? I can't see any CI Test being run on Docker profile for this PR.

Is this because the merge is between my feature branch and the public_dev branch rather than being to the main dev branch?

**Comment by gq1** [comment]
Every time I start the pipeline, there is a warning message not sure which line yet. Can we remove it?
`WARN: The operator `first` is useless when applied to a value channel which returns a single value by definition`

**Comment by gq1** [comment]
I tried `nextflow run genomenote/main.nf -profile singularity,sanger,test` on tol farm 3 times.

`FASTK_FASTK` always failed in the first attempt and stuck on the second attempt.

Because this PR is mainly for higlass loading. Tried the following again:
`nextflow run genomenote/main.nf -profile singularity,sanger,test -resume --upload_higlass_data --higlass_data_project_dir /asg/algae_gq2 --higlass_kubeconfig /nfs/users/nfs_g/gq2/tol-it-dev-config-openlab` 

The job again was stuck on `FASTK_FASTK` but passed the higlass_loading step successfully.
I can find the output file from the work directory with higlass link, `uoEpiScrs1_T3_higlass_link.csv` 
http://genome-note-higlass.tol-dev.sanger.ac.uk/l/?d=GCA_946965045_2_uoEpiScrs1_T3

And the link works and I also can see the project and tilesets I created on higlass server.

Screenshot 2024-02-02 at 11 31 37

**Comment by gq1** [comment]
@BethYates can run the test without `sanger` profile. I tried `nextflow run genomenote/main.nf -profile singularity,test`

failed on `SANGERTOL_GENOMENOTE:GENOMENOTE:CONTACT_MAPS:SAMTOOLS_VIEW (uoEpiScrs1_T3)`

with the following error message:
```
Caused by:
  Process requirement exceeds available CPUs -- req: 2; avail: 1
```

File: bin/generate_higlass_link.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python3
```
Comment (gq1) [inline bin/generate_higlass_link.py:1]: This is probably a habit, I like to have some comments to explain the script a bit.

**Comment by gq1** [comment]
I tried `nextflow run genomenote/main.nf -profile singularity,sanger,test -resume --upload_higlass_data --higlass_data_project_dir /asg/algae_gq2 --higlass_kubeconfig /nfs/users/nfs_g/gq2/tol-it-dev-config-openlab` again.

I can see the old tilesets being deleted from the server and the file systems. And then same link being generated with the new tilesets.

**Comment by BethYates** [comment]
> Every time I start the pipeline, there is a warning message not sure which line yet. Can we remove it? `WARN: The operator `first` is useless when applied to a value channel which returns a single value by definition`

I'll address this when I do the merge from the dev branch to the public dev branch after the PR is merged

File: modules/local/upload_higlass_data.nf [hunk 2]
```diff
@@ -45,10 +50,37 @@ process UPLOAD_HIGLASS_DATA {
     cp -f $genome ${upload_dir}${project_name}/${file_name}.genome
 
     # Load them in Kubernetes
+
```
[the remark this answers is not in the document]
Comment (BethYates) [reply modules/local/upload_higlass_data.nf:53]: I've created a script upload_higlass_file.sh

**Pull request merged**